pkg/log/zap/flags.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package zap contains helpers for setting up a new logr.Logger instance
+// using the Zap logging framework.
+
+package zap
+
+import (
+	"fmt"
+	"strings"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+type encoderConfigFunc func(*zapcore.EncoderConfig)
+
+type encoderValue struct {
+	set     bool
+	setFunc func(zapcore.Encoder)
+	str     string
+}
+
+func (v encoderValue) String() string {
+	return v.str
+}
+
+func (v encoderValue) Type() string {
+	return "encoder"
+}
+
+func (v *encoderValue) Set(e string) error {
+	v.set = true
+	val := strings.ToLower(e)
+	switch val {
+	case "json":
+		v.setFunc(newJSONEncoder())
+	case "console":
+		v.setFunc(newConsoleEncoder())
+	default:
+		return fmt.Errorf("invalid encoder value \"%s\"", e)
+	}
+	v.str = e
+	return nil
+}
+
+func newJSONEncoder(ecfs ...encoderConfigFunc) zapcore.Encoder {
+	encoderConfig := zap.NewProductionEncoderConfig()
+	for _, f := range ecfs {
+		f(&encoderConfig)
+	}
+	return zapcore.NewJSONEncoder(encoderConfig)
+}
+
+func newConsoleEncoder(ecfs ...encoderConfigFunc) zapcore.Encoder {
+	encoderConfig := zap.NewDevelopmentEncoderConfig()
+	for _, f := range ecfs {
+		f(&encoderConfig)
+	}
+	return zapcore.NewConsoleEncoder(encoderConfig)
+}
+
+type levelValue struct {
+	set     bool
+	setFunc func(zap.AtomicLevel)
+	str     string
+}
+
+func (v *levelValue) Set(l string) error {
+	v.set = true
```

Comment:
@joelanford Is `set` actually needed? 
E.g we're not currently checking this to see if the flag has already been set once.
And I don't see a method in the [Value](https://github.com/spf13/pflag/blob/HEAD/flag.go#L187:6) interface using this as well.

Reply:
Good catch. That's leftover from how we were doing defaulting in SDK (e.g. `if !set { v = defaultVal }`)

Here, we're just calling the setFunc and then deferring to the existing `addDefaults` function.

So I agree. I don't think `set` is needed with this implementation.